benchmark/spmv/cuda_linops.hpp

```diff
@@ -0,0 +1,440 @@
+
+public:
+    using csr = gko::matrix::Csr<ValueType, IndexType>;
+    using mat_data = gko::matrix_data<ValueType, IndexType>;
+
+    void read(const mat_data &data) override
+    {
+        Csr_->read(data);
+        this->set_size(gko::dim<2>{Csr_->get_size()});
+    }
+
+    void apply_impl(const gko::LinOp *b, gko::LinOp *x) const override
+    {
+        auto dense_b = gko::as<gko::matrix::Dense<double>>(b);
+        auto dense_x = gko::as<gko::matrix::Dense<ValueType>>(x);
+        auto db = dense_b->get_const_values();
+        auto dx = dense_x->get_values();
+        auto alpha = gko::one<ValueType>();
+        auto beta = gko::zero<ValueType>();
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseDcsrmv_mp(
+            this->get_gpu_exec()->get_cusparse_handle(), trans_,
+            this->get_size()[0], this->get_size()[1],
+            Csr_->get_num_stored_elements(), &alpha, this->get_descr(),
+            Csr_->get_const_values(), Csr_->get_const_row_ptrs(),
+            Csr_->get_const_col_idxs(), db, &beta, dx));
+    }
+
+    gko::size_type get_num_stored_elements() const noexcept
+    {
+        return Csr_->get_num_stored_elements();
+    }
+
+    CuspCsrmp(std::shared_ptr<const gko::Executor> exec,
+              const gko::dim<2> &size = gko::dim<2>{})
+        : gko::EnableLinOp<CuspCsrmp, CuspBase>(exec, size),
+          Csr_(std::move(
+              csr::create(exec, std::make_shared<typename csr::classical>()))),
+          trans_(CUSPARSE_OPERATION_NON_TRANSPOSE)
+    {}
+
+
+private:
+    std::shared_ptr<csr> Csr_;
+    cusparseOperation_t trans_;
+};
+
+
+template <typename ValueType = gko::default_precision,
+          typename IndexType = gko::int32>
+class CuspCsr
+    : public gko::EnableLinOp<CuspCsr<ValueType, IndexType>, CuspBase>,
+      public gko::EnableCreateMethod<CuspCsr<ValueType, IndexType>>,
+      public gko::ReadableFromMatrixData<ValueType, IndexType> {
+    friend class gko::EnableCreateMethod<CuspCsr>;
+
+public:
+    using csr = gko::matrix::Csr<ValueType, IndexType>;
+    using mat_data = gko::matrix_data<ValueType, IndexType>;
+
+    void read(const mat_data &data) override
+    {
+        Csr_->read(data);
+        this->set_size(gko::dim<2>{Csr_->get_size()});
+    }
+
+    void apply_impl(const gko::LinOp *b, gko::LinOp *x) const override
+    {
+        auto dense_b = gko::as<gko::matrix::Dense<ValueType>>(b);
+        auto dense_x = gko::as<gko::matrix::Dense<ValueType>>(x);
+        auto db = dense_b->get_const_values();
+        auto dx = dense_x->get_values();
+        auto alpha = gko::one<ValueType>();
+        auto beta = gko::zero<ValueType>();
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseDcsrmv(
+            this->get_gpu_exec()->get_cusparse_handle(), trans_,
+            this->get_size()[0], this->get_size()[1],
+            Csr_->get_num_stored_elements(), &alpha, this->get_descr(),
+            Csr_->get_const_values(), Csr_->get_const_row_ptrs(),
```

Comment:
Does the `values` array need to passed through [`as_cuda_type`](https://github.com/ginkgo-project/ginkgo/blob/2c7b1ed0a6cda84e32f7a238d9320c9fdf84489b/cuda/base/types.hpp#L154) ?

Reply:
They probably should. That's a good question though as it somehow works as such.

Reply:
No in fact they do not have to, because CUDA does that itself. We need to do this in our case only because we have our own kernel calls, and inside the actual kernels a double needs to become the CUDA specific double version.

When you look at the prototype, they use the data either as `double *` or `void *` with a type flag. See for example the following:
https://docs.nvidia.com/cuda/cusparse/#cusparse-csrmvEx
https://docs.nvidia.com/cuda/cusparse/#cusparse-lt-t-gt-csrmv_mergepath

Reply:
Ah okay. That makes sense.